### ISSUE TITLE
TestContext does not flow in async method

### ISSUE THREAD

**Issue by Suremaker** [nunit/nunit#1965]
Example tests recreating issue:
```C#
[TestFixture]
public class TestContext_async_issue
{
    [Test]
    public async Task TestContext_Out_should_flow_with_async_execution()
    {
        var expected = TestContext.Out;
        await Task.Yield();
        Assert.AreSame(expected, TestContext.Out);
    }

    [Test]
    public async Task TestContext_CurrentContext_should_flow_with_async_execution()
    {
        var expected = TestContext.CurrentContext;
        await Task.Yield();
        Assert.AreSame(expected, TestContext.CurrentContext);
    }
}
```

The side effect of that behaviour is that it is not possible to use `TestContext.WriteLine()` after await as it causes `NullReferenceException`.

Tested with `NUnit 3.5.0`, `net46` and `netcoreapp1.0`

**Comment by Suremaker** [comment]
The problem is that `TestExecutionContext` is holding current context with `[ThreadStatic]` attribute ([source code link](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/TestExecutionContext.cs#L182)).

While it works fine in scenarios where multiple tests are executed in parallel (but each in given thread at a time), in async scenarios it will not work as each method continuation may execute in other thread.
Replacement of `[ThreadStatic]` with [AsyncLocal<T>](https://msdn.microsoft.com/en-us/library/dn906268(v=vs.110).aspx) would be solution here, but unfortunately this class is available since .net 4.6 which is a pain for frameworks supporting older net versions :(

**Comment by rprouse** [comment]
Thanks for the analysis. It doesn't sound like it is going to be an easy fix.

**Comment by Suremaker** [comment]
For pre 4.6 there are some workarounds like:
* [CallContext](http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html) however it would not work below 4.5 (according to article) as well as should be replaced with AsyncLocal in >4.6
* custom implementation of [SynchronizationContext](https://msdn.microsoft.com/en-us/library/system.threading.synchronizationcontext(v=vs.110).aspx) however implementing it right in pre 4.5 will be difficult, as way how tasks were supporting SynchronizationContext flow was buggy - it is available since .net 2.0 however.

**Comment by jhickson** [comment]
Could the issue I reported a few days ago in #1952 be explained by this? Though there .net46 is fine - it's just .netcoreapp1.1 that is the issue.

**Comment by Suremaker** [comment]
@jhickson I think those issues are correlated to some extent. Definitely in your scenario task is being executed in another thread, so it will not share the context that has been setup for test method.

I am not sure however if fixing async/await scenarios will make your scenario working as well, as you are using TaskFactory explicitly and I do not know if it would be enough in this case.

**Comment by CharliePoole** [comment]
They are both part of a more general problem: TestExecutionContext doesn't get properly initialized if the test creates a thread outside of NUnit's control. Issue #1952 is a perfect example.

This is as well, except that there's some chance we can get control of the thread creation by using our own SynchronizationContext.

@Suremaker's analysis is correct as far as the Portable and NetStandard builds go. I used ThreadStatic in Portable because it's the only thing I could find. The desktop builds use CallContext and the now removed CF build did something else.

Since the implementations are different in each build, it's important that any comments on this problem report which build is in use.

**Comment by Suremaker** [comment]
Hello,

I made a commit on my fork that solves the issue for NetStandard1.6: https://github.com/Suremaker/nunit/commit/4349f80e779213be34ed38ab6c4f47496c64f234
I imagine that similar change should be done for NET46 and above but such do not exist yet in the solution so I did not go that far.

I modified the tests slightly because I realized that second original test will always fail, so now they are as follows:
```c#
         [Test]
         public async Task TestContext_Out_should_flow_with_async_execution()
         {
             var expected = TestContext.Out;
             await Task.Yield();
             Assert.AreSame(expected, TestContext.Out);
         }

         [Test]
         public async Task TestExecutionContext_CurrentContext_should_flow_with_async_execution()
         {
             var expected = TestExecutionContext.CurrentContext;
             await Task.Yield();
             Assert.AreSame(expected, TestExecutionContext.CurrentContext);
         }
```

Later I will check if this change will cover the issue with `netcoreapp1.0` as right now I really do not know if in this scenario a portable (broken) version is used or netstandard1.6.

When proven working and you would be happy with solution I can make a pull request.

**Comment by rprouse** [comment]
#1973 fixed this for .NET Standard

**Comment by Suremaker** [comment]
I wanted to check if there is the new release date know for this fix yet?

**Comment by CharliePoole** [comment]
The merged PR fixes the problem for .NET Standard only. The issue is still open with respect to all other platforms. We would welcome people who use .NET Standard trying it, by getting the latest development build from our MyGet feed at https://www.myget.org/F/nunit/api/v2. Please let us know if it works for you!

In the normal course of events, this fix will be in the next release. Releases are scheduled quarterly and the next one (3.7) is due on March 10 - actually a bit earlier than the end of the quarter. For critical fixes, we do interim releases (3.6.1, 3.6.2, etc.) but we aren't treating this one as critical.

Since .NET standard support is still somewhat experimental, I'd advise you to use a dev release if you are dependent on it. I wouldn't suggest doing automatic updates of those pre-releases, since there could be a lot of thrashing involved, but keep track of what fixes are merged and update when it's useful.

**Comment by jnm2** [comment]
Yay, I just hit this with net452. There's no chance we'll be able to use netstandard soon.

What about a null check in case you're asserting without a test context? That might tide everyone over.
https://github.com/nunit/nunit/blob/8ae0489f20c96bffb3a720214a07f66f7e503dcc/src/NUnitFramework/framework/Assert.cs#L364-L368

**Comment by CharliePoole** [comment]
@jnm2 I don't understand what you're using that's giving the error. Is it the latest release, a build from master or a dev release?

**Comment by jnm2** [comment]
Sorry, it's the latest release but I didn't notice a critical detail here... I was running via a ReSharper ad-hoc entry point so that I could press F5 and by test would run. So of course the test context is null, because NUnit isn't in the picture until I get to the assert. This is really a separate issue: supporting asserts with no context. It would be super helpful if it just worked.

**Comment by CharliePoole** [comment]
This has never been a supported way of running tests, although I understand that it has worked for you in the past. In any case, it relies on NUnit's using an exception mechanism to report failures, which has always been an implementation detail subject to change. With the latest release, we have made a start on that change and in subsequent releases, it's possible that future releases will only use the exception mechanism to terminate the test, not to report the error.

**Comment by jnm2** [comment]
I figured. Maybe I can use NUnitLite for this.

**Comment by CharliePoole** [comment]
@Suremaker I believe this is now fixed in master. Can you try one of our myget builds, NUnit 3.7.0-dev-03707 or greater?

**Comment by Suremaker** [comment]
@CharliePoole I have rerun my tests with NUnit 3.7.0-dev-03707 and they passed now (while previously they were failing with NUnit 3.6.0).

I am looking forward to seeing official version of NUnit 3.7.
Thanks!

**Comment by CharliePoole** [comment]
Good to hear! We expect to put out 3.7 toward the end of March.

**Comment by SamuelCox** [comment]
Hi, is there any chance this is still an issue? I'm getting something similar, that I initially logged with the corefx repo, but given the error only occurs in a unit test, and not just normal code, perhaps I should have logged here first.

https://github.com/dotnet/corefx/issues/26206

**Comment by jnm2** [comment]
It's a bit hard to follow since the call stack says it happened on line 24. Are you saying that `calendarService.CreateEvent(calendarEventModel);` returned null and then the `await` caused the NRE?

**Comment by jnm2** [comment]
My bad, found the location: https://github.com/SamuelCox/LifeManager/blob/master/LifeManager/LifeManager.CalendarService/Services/CalendarService.cs#L24

**Comment by SamuelCox** [comment]
I'll try downgrading to a lower version of nunit when I have a moment, but yes that is the location

**Comment by jnm2** [comment]
What was null, `_repository`, or the return value of `_repository.Add`?

**Comment by SamuelCox** [comment]
The return value of _repository.Add

**Comment by jnm2** [comment]
I'd make the constructor throw ArgumentNullException btw for `repo`.

**Comment by SamuelCox** [comment]
For what it's worth, I got the same behaviour when I hacked the repository add method to just return a Task<bool> and return true

**Comment by jnm2** [comment]
I'm getting the same behavior without NUnit being invoked. Add this class:

```cs

namespace LifeManager.CalendarService.Tests
{
    public static class Program
    {
        public static void Main()
        {
            var fixture = new CalendarServiceTests();
            fixture.SetUp();
            fixture.CreateEvent_ShouldCallRepositoryAdd().GetAwaiter().GetResult();
        }
    }
```

And overwrite these csproj proerties:

```xml
    <OutputType>Exe</OutputType>
    <StartupObject>LifeManager.CalendarService.Tests.Program</StartupObject>
```

And press F5.

**Comment by jnm2** [comment]
In fact, awaits don't have anything to do with it either as you can see by running this:

```cs
public static void Main()
{
    Mapper.Initialize(cfg => cfg.CreateMap<CalendarEvent, CalendarEventModel>().ReverseMap());
    var mockCalendarRepository = new Mock<ICalendarRepository>();
    mockCalendarRepository.Setup(x => x.Add(It.IsAny<CalendarEvent>())).Verifiable();

    var calendarEventModel = new CalendarEventModel { Id = Guid.NewGuid() };
    var calendarEvent = Mapper.Map<CalendarEvent>(calendarEventModel);
    Assert.That(mockCalendarRepository.Object.Add(calendarEvent), Is.Not.Null);
}
```

**Comment by jnm2** [comment]
Looking closer, it seems that you haven't set up the mock to return anything in particular when you call Add on the mock instance. So returning null is what I'd expect it to do.

**Comment by SamuelCox** [comment]
Yeah, no, you're right. That's my inexperience with async await, as I was still considering the call essentially a void call, so wasn't thinking I'd need to mock a return value for that. Thanks for this. Sorry for the noise on this issue

**Comment by jnm2** [comment]
To keep this thread on topic, I've replied at https://github.com/dotnet/corefx/issues/26206#issuecomment-355792422.

**Comment by jnm2** [comment]
Happy to help! It might make sense for Moq to default to returning `Task.CompletedTask` rather than `null`. That could be a good feature request.
